Add a new article on effects (#1)
fagossa/fagossa.github.io: _posts/2019-12-17-effects.md

---
layout: post
title:  "Common techniques with cats-effects"
categories: speaker
---

# Cats-effect

> The goal of this article is to document some _cats-effect_ techniques that could be helpful to start working with effets.

Reviewer: Little typo -> Effects

## Defining our domain model

> We are going to experiment with an application that will request the temperature of a give city, keep the history of the requests and finally show the hottest city so far.

Reviewer: Little typo -> given

```scala
object domain {

  // the location of an imaginary external forecast service
  case class ForecastConfig(host: String, port: Int)

  sealed trait TempUnit extends Product with Serializable
  case object Celcius extends TempUnit

  case class Temperature(value: Int, unit: TempUnit = Celcius)

  case class City(name: String)

  case class Forecast(temperature: Temperature)

  // Our business errors
  sealed trait AppErrors extends NoStackTrace with Product with Serializable
  case class UnknownCity(name: String) extends AppErrors

}
```

## Writing to the console
As we are going to develop a console application we are going to need a way to
wrap thoses interactions with effects. The simplest behavior is  to
read and write from the terminal:

```scala
trait Console[F[_]] {
  def printLn(line: String): F[Unit]
  def readLn: F[String]
}
```
In order to use an implementation of this _trait_ we have two alternatives:
* Add an _implicit_ parameter, `def hello[F[_]](implicit C: Console[F])`
* Add bounds to `F[_]`, for example `def hello[F[_] : Console]`

> The former gets really verbose and the later require an `implicitly` every single time: `implicitly[Console[F]].printLn("hello")`. I personally prefer the second one with a simple bonus: hiden that implicit call behind an `apply`:

Reviewer: -> To hide that implicit

```scala
object Console {
  def apply[F[_]: Console]: Console[F] = implicitly[Console[F]]
}
```

> So,  every time we need to write to the console, we just use:

Reviewer: remove extra space.

```scala
object domain {

  // the location of an imaginary external forecast service
  case class ForecastConfig(host: String, port: Int)

  sealed trait TempUnit extends Product with Serializable
  case object Celcius extends TempUnit

  case class Temperature(value: Int, unit: TempUnit = Celcius)

  case class City(name: String)

  case class Forecast(temperature: Temperature)

  // Our business errors
  sealed trait AppErrors extends NoStackTrace with Product with Serializable
  case class UnknownCity(name: String) extends AppErrors

}
```

## Writing to the console
As we are going to develop a console application we are going to need a way to
wrap thoses interactions with effects. The simplest behavior is  to
read and write from the terminal:

```scala
trait Console[F[_]] {
  def printLn(line: String): F[Unit]
  def readLn: F[String]
}
```

In order to use an implementation of this _trait_ we have two alternatives:
* Add an _implicit_ parameter, `def hello[F[_]](implicit C: Console[F])`
* Add bounds to `F[_]`, for example `def hello[F[_] : Console]`

> The former gets really verbose and the later require an `implicitly` every single time: `implicitly[Console[F]].printLn("hello")`. I personally prefer the later with a simple bonus: hide that implicit call behind an `apply`:

Reviewer: Typo -> requires

> The former gets really verbose and the later require an `implicitly` every single time: `implicitly[Console[F]].printLn("hello")`. I personally prefer the later with a simple bonus: hide that implicit call behind an `apply`:

Reviewer: every single time -> each time  ?

```scala
object Console {
  def apply[F[_]: Console]: Console[F] = implicitly[Console[F]]
}
```
So,  every time we need to write to the console, we just use:

> `Console[F]].printLn("hello")`.

Reviewer: - It feels like there is a missing step here. Something like defining an interpreter for Console algebra, and you can use it on the example above. (Console[F]].printLn("hello")`)

Another alternative is to use [console4cats](https://github.com/profunktor/console4cats) that implements this for us.

## Business logic

### The weather client
The default implementation of our client has some hardcoded values.
Our _trait_ has only one method, that means that we can easilly transform our implementation into an _Single Abstract Method_ (SAM).

> Notice how our companion object has smart constructors pour each implementation of the client.

Reviewer: pour each implementation ^^ -> for each implementation

```scala
trait WeatherClient[F[_]]{
  def forecast(city: City): F[Forecast]
}
object WeatherClient {

>   def make[F[_]: Sync](host: String, port: Int): F[WeatherClient[F]] = Sync[F].pure {

Reviewer: is it normal that host & port are not used at this point ?

Author: Oui, c'est juste pour l'exemple

    case City("Paris") => Sync[F].pure(Forecast(Temperature(28)))
    case City("Lyon") => Sync[F].pure(Forecast(Temperature(34)))
    case _ => Sync[F].pure(Forecast(Temperature(15)))
  }
}
```

> In order to build the client, we need to extract certains values from the configuration. We could do this using the MTL approach with a `ApplicativeAsk` but our performance would be (severily) penalised. A much simpler approach is to specify the configuration as a parameter and then, extract the data as _effects_ (we could do some validations as required).

Reviewer: certains values -> some values

Reviewer: penalised -> penalized

Reviewer: with a 'ApplicativeAsk' -> with an 'ApplicativeAsk'

```scala
def host[F[_]: Sync](config: ForecastConfig): F[String] = Sync[F].pure(config.host)

def port[F[_]: Sync](config: ForecastConfig): F[Int] = Sync[F].pure(config.port)
```

### Handling state
The history of the requests is stored in a [Ref](https://typelevel.org/cats-effect/concurrency/ref.html). A `Ref` is basically a _thread safe_ container. Here, we just get the current state, sort by temperature and get the first element.

```scala
type Requests = Map[City, Forecast]
import cats.effect.concurrent.Ref

def hottestCity[F[_]: Functor](
  ref: Ref[F, Requests]
): F[Option[(City, Temperature)]] = {
  import cats.syntax.functor._
  ref.get.map(reqs =>
    reqs.toList
      .sortBy(_._2.temperature.value)
      .map {

>         case (city, forcast) => (city, forcast.temperature)

Reviewer: typo -> forecast

      }
      .reverse
      .headOption)
}
```

### Handling interactions with the user

> Demanding a city is basically printing a message and waiting for the used to type the city name

Reviewer: ```suggestion
Asking to provide a city is essentially about printing a message and waiting for the user to type in the city's name. 
```

```scala
def askCity[F[_]: Console: Monad]: F[String] = {
  import cats.syntax.functor._, cats.syntax.flatMap._
  for {
    _        <- Console[F].printLn("What is your next city?")
    cityName <- Console[F].readLn
  } yield cityName
}
```

### Handling errors

> When we want to avoid polluting all signatures with `Option[T]` or `Either[Error, T]`, we could use an implicit `MonadError[F]` which is already provider by `Sync[T]`.

Reviewer: provider -> provided

  override def forecast(city: City): F[Forecast] = Sync[F].pure {
    case City("Paris") => Sync[F].pure(Forecast(Temperature(28)))
    case City("Lyon") => Sync[F].pure(Forecast(Temperature(34)))
    case _ => Sync[F].pure(Forecast(Temperature(15)))
  }
}
```
Notice how our companion object has a smart constructor to build the actual implementation of the client. Also we want to hice the constructor to avoid leaking the state of our implementation.

### Handling state
The history of the requests is stored in a [Ref](https://typelevel.org/cats-effect/concurrency/ref.html). A `Ref` is basically an _Atomic Reference_ that can show itself as a _State Monad_. Here, we just get the current state, sort by temperature and get the first element.

```scala
type Requests = Map[City, Forecast]
import cats.effect.concurrent.Ref

def hottestCity[F[_]: Functor](
  ref: Ref[F, Requests]
): F[Option[(City, Temperature)]] = {
  import cats.syntax.functor._
  ref.get.map(reqs =>
    reqs.toList
      .sortBy(_._2.temperature.value)
      .map {
        case (city, forcast) => (city, forcast.temperature)
      }
      .reverse
      .headOption)
}
```

### Handling interactions with the user

> Asking for a city is basically printing a message and waiting for the used to type the city name.

Reviewer: Used -> User

and repeats the main loop forever.
```scala
def program[F[_]: Console: Sync](
  config: ForecastConfig
): F[Unit] = {
  import cats.syntax.functor._, cats.syntax.flatMap._
  for {
    h      <- host[F](config)
    p      <- port[F](config)
    client <- LiveWeatherClient.make(h, p)
    _      <- Console[F].printLn(s"Using weather service at http://$h:$p \n")
    state  <- Ref.of[F, Map[City, Forecast]](Map.empty[City, Forecast])
    _      <- askLoop[F](client, state).foreverM[Unit]
  } yield ()
}
```

### Running our application
Finally, to run our application we should call the program and provide an interpreter of _Console_ for _IO_.

```scala
object ForecastApp extends IOApp {
  // this is the interpreter for Console
  implicit val ioConsole: Console[IO] = new Console[IO] {
    override def printLn(line: String): IO[Unit] = IO.delay(println(line))
    override def readLn: IO[String] = IO.delay(scala.io.StdIn.readLine())
  }

  override def run(args: List[String]): IO[ExitCode] = {
    for {
      config <- IO.pure(domain.ForecastConfig("localhost", 8080))
      _      <- Logic.program[IO](config)
    } yield ExitCode.Success
  }
}
```

## Conclusions

> TODO

Reviewer 2: a venir mais je pense qu'il faudrait un bon petit résumé des différentes techniques
